hasktorch/src/Torch/Static/Native.hs

```diff
@@ -22,20 +22,1224 @@ import Data.Proxy
 import Data.Finite
 import Data.Kind (Constraint)
 import GHC.TypeLits
+import qualified Data.Int as I
 
-import Prelude hiding (sin)
+import System.IO.Unsafe
+import Foreign.ForeignPtr
+
+import qualified ATen.Managed.Native as ATen
+import qualified ATen.Managed.Type.Tensor as ATen
+import qualified ATen.Managed.Type.Scalar as ATen
+import qualified ATen.Managed.Type.Tuple as ATen
+import qualified ATen.Const as ATen
+import qualified ATen.Type as ATen
+import qualified ATen.Managed.Cast
+import ATen.Cast
+
+
+import Prelude hiding (sin, abs, max, min)
 import qualified Torch.Tensor as D
 import qualified Torch.TensorFactories as D
-import Torch.Functions as D
+import Torch.Functions (Reduction(..), Tri(..), isUpper, kOne)
 import Torch.DType
 import Torch.Static
+import Torch.Static.Factories
+import Torch.Scalar
 import qualified Torch.TensorOptions as D
 import Data.Reflection
+import ATen.Cast
 
-sin :: Tensor dtype shape -> Tensor dtype shape
-sin t = UnsafeMkTensor $ D.sin (toDynamic t)
+---
+
+dim :: Tensor dtype shape -> Int
+dim t = D.dim $ toDynamic t
+
+shape :: Tensor dtype shape -> [Int]
+shape t = D.shape $ toDynamic t
+
+dtype :: Tensor dtype shape -> DType
+dtype t = D.dtype $ toDynamic t
+
+toInt :: Tensor dtype shape -> Int
+toInt t = D.toInt $ toDynamic t
+
+sumAll :: Tensor dtype shape -> Tensor dtype shape
+sumAll t = unsafePerformIO $ (cast1 ATen.sum_t) t
+
+-- |
+-- >>> t = abs (ones :: Tensor Float '[2,2])
+-- >>> (dtype t,shape t)
+-- (Float,[2,2])
+abs :: Tensor dtype shape -> Tensor dtype shape
+abs t = unsafePerformIO $ (cast1 ATen.abs_t) t
+
+-- add :: Tensor dtype shape -> Tensor dtype shape -> Tensor dtype shape
+-- add a b = unsafePerformIO $ (cast3 ATen.add_tts) a b kOne
+
+ceil :: Tensor dtype shape -> Tensor dtype shape
+ceil t = unsafePerformIO $ (cast1 ATen.ceil_t) t
+
+floor :: Tensor dtype shape -> Tensor dtype shape
+floor t = unsafePerformIO $ (cast1 ATen.floor_t) t
+
+-- |
+-- >>> t = min (ones :: Tensor Float '[2,2])
+-- >>> (dtype t,shape t)
+-- (Float,[])
+min :: Tensor dtype shape -> Tensor dtype '[]
+min t = unsafePerformIO $ (cast1 ATen.min_t) t
+
+-- |
+-- >>> t = max (ones :: Tensor Float '[2,2])
+-- >>> (dtype t,shape t)
+-- (Float,[])
+max :: Tensor dtype shape -> Tensor dtype '[]
```

Comment:
nit: I'd keep `max` for the element-wise operation and call the reduction operators `maxAll` and `maxDim`

Reply:
Thank you for your feedback.

Add {max,min,median}Dim 
https://github.com/hasktorch/hasktorch/pull/163/files#diff-ed5b0b77220a9f439e61aa413e33bdfdR1257-R1270